<?xml version="1.0"?>
<robot name="snakebot">

  <link name="base_link">
    <visual>
     <geometry>
        <cylinder length="0.6" radius="0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </visual>
    <collision>
      <geometry>
      <cylinder length="0.6" radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>

  <link name="link_1">
    <visual>
      <geometry>
        <cylinder length="0.6" radius="0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.3"/>
   </visual>
    <collision>
      <geometry>
      <cylinder length="0.6" radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>

  <joint name="joint_1" type="revolute">
    <parent link="base_link"/>
    <child link="link_1"/>
    <axis xyz= "0 1 0"/>
    <origin xyz="0 0 0.3"/>
    <limit effort="1000" velocity="1.0" lower="-0.45" upper="0.45" />
  </joint>

  <link name="link_2">
    <visual>
      <geometry>
        <cylinder length="0.6" radius="0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.3"/>
   </visual>
    <collision>
      <geometry>
      <cylinder length="0.6" radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>

  <joint name="joint_2" type="revolute">
    <parent link="link_1"/>
    <child link="link_2"/>
    <axis xyz= "1 0 0"/>
    <origin xyz="0 0 0.6"/>
    <limit effort="1000" velocity="1.0" lower="-0.45" upper="0.45" />
  </joint>

  <link name="link_3">
    <visual>
      <geometry>
        <cylinder length="0.6" radius="0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.3"/>
   </visual>
    <collision>
      <geometry>
      <cylinder length="0.6" radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>

  <joint name="joint_3" type="revolute">
    <parent link="link_2"/>
    <child link="link_3"/>
    <axis xyz= "0 1 0"/>
    <origin xyz="0 0 0.6"/>
    <limit effort="1000" velocity="1.0" lower="-0.45" upper="0.45" />
  </joint>

  <link name="link_4">
    <visual>
      <geometry>
        <cylinder length="0.6" radius="0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.3"/>
   </visual>
    <collision>
      <geometry>
      <cylinder length="0.6" radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>

  <joint name="joint_4" type="revolute">
    <parent link="link_3"/>
    <child link="link_4"/>
    <axis xyz= "1 0 0"/>
    <origin xyz="0 0 0.6"/>
    <limit effort="1000" velocity="1.0" lower="-0.45" upper="0.45" />
  </joint>

  <link name="link_5">
    <visual>
      <geometry>
        <cylinder length="0.6" radius="0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.3"/>
   </visual>
    <collision>
      <geometry>
      <cylinder length="0.6" radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>

  <joint name="joint_5" type="revolute">
    <parent link="link_4"/>
    <child link="link_5"/>
    <axis xyz= "0 1 0"/>
    <origin xyz="0 0 0.6"/>
    <limit effort="1000" velocity="1.0" lower="-0.45" upper="0.45" />
  </joint>

  <link name="link_6">
    <visual>
      <geometry>
        <cylinder length="0.6" radius="0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.3"/>
   </visual>
    <collision>
      <geometry>
      <cylinder length="0.6" radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>

  <joint name="joint_6" type="revolute">
    <parent link="link_5"/>
    <child link="link_6"/>
    <axis xyz= "1 0 0"/>
    <origin xyz="0 0 0.6"/>
    <limit effort="1000" velocity="1.0" lower="-0.45" upper="0.45" />
  </joint>

  <link name="link_7">
    <visual>
      <geometry>
        <cylinder length="0.8" radius="0.2"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0.3"/>
   </visual>
    <collision>
      <geometry>
      <cylinder length="0.6" radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>

  <joint name="joint_7" type="revolute">
    <parent link="link_6"/>
    <child link="link_7"/>
    <axis xyz= "0 1 0"/>
    <origin xyz="0 0 0.6"/>
    <limit effort="1000" velocity="1.0" lower="-0.45" upper="0.45" />
  </joint>




  <!-- ROS Control plugin for Gazebo -->
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/snakebot</robotNamespace>
    </plugin>
  </gazebo>

  <!-- transmission -->
  <transmission name="joint_1_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor1">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint_1">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="joint_2_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor2">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint_2">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="joint_3_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor3">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint_3">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="joint_4_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor4">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint_4">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="joint_5_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor5">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint_5">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>

  <transmission name="joint_6_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor6">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint_6">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>


  <transmission name="joint_7_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor7">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint_7">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>


</robot>
